### ISSUE TITLE
Add support for drop!/reword! commits to `git rebase -i`

### ISSUE THREAD

**Issue by dscho** [gitgitgadget/git#259]
This was already discussed briefly on the Git mailing list: https://public-inbox.org/git/alpine.DEB.2.21.1.1710151754070.40514@virtualbox/

In short, the reword! semantics are relatively easy in my mind: it would be called as

	git commit --reword <commit>

would take staged changes (if any), otherwise implicitly --allow-empty,
then create a commit message to be edited in this style:

	reword! <original-oneline>

	<original-oneline>

	<original-body>

This would be presented to the user in an editor (similar to `git commit
--amend`).

Upon rebase --autosquash, it would work like a squash! but comment out all
previous commit messages, and also comment out the reword! line and the
empty line after that.

In case of multiple reword!, the last one would win, and it would
naturally integrate in any fixup!/squash! workflow.

What is *more* difficult is something else I frequently need: drop!
<oneline>. That is, I want to explicitly mark a commit to be excluded in
subsequent rebase --autosquash. I *guess* the best way would be to
implement a

	git revert --drop <commit>

that would work as if you called `git revert -n <commit> && git commit -m
'drop! '"$(git show -s --oneline <commit>)", and upon rebase --autosquash,
it would reorder like fixup!/squash!/reword!, replace the `pick` of the
previous line (if it was a `pick`) by `drop` and comment out the current
`pick drop! <oneline>` line.

The reason why the semantics are more difficult in that case is that drop!
does not mix well with fixup!/squash!/reword!.

**Comment by Hasaber8** [comment]
Has this project already been taken up or is still available to be worked upon?
-Rohan

**Comment by dscho** [comment]
I am unaware of any effort so far, so go right ahead!

**Comment by Hasaber8** [comment]
Correct me if I'm wrong , current even if we try and skip a single commit in git rebase -i it gets squashed, without it been shown in th editor.

**Comment by Hasaber8** [comment]
And the whole point of doing to is be able to keep the commit intact while being able to squash commits around it.

**Comment by dscho** [comment]
> Correct me if I'm wrong , current even if we try and skip a single commit in git rebase -i it gets squashed, without it been shown in th editor.

If I understand your question right, you're asking whether `git rebase --skip` still applies the currently-picked commit? It does not...

**Comment by Hasaber8** [comment]
> > Correct me if I'm wrong , current even if we try and skip a single commit in git rebase -i it gets squashed, without it been shown in th editor.
> 
> If I understand your question right, you're asking whether `git rebase --skip` still applies the currently-picked commit? It does not...

The drop! semantics if executed properly will be a very good thing to use! This surely will be an interesting project to do!

**Comment by phillipwood** [comment]
> ```
> git revert --drop <commit>
> ```
> 
> that would work as if you called `git revert -n <commit> && git commit -m 'drop! '"$(git show -s --oneline <commit>)"`,

Do we need to actually do the revert? - If we want to drop the commit can't we use an empty commit (`git commit -m 'drop! '"$(git show -s --oneline <commit>)"`)? If we actually try to revert the commit there is the potential for conflicts when running revert which would create unnecessary work for the user.

**Comment by dscho** [comment]
> > ```
> > git revert --drop <commit>
> > ```
> > 
> > 
> > that would work as if you called `git revert -n <commit> && git commit -m 'drop! '"$(git show -s --oneline <commit>)"`,
> 
> 
> Do we need to actually do the revert?

Yes. The point of dropping a given commit is to drop its changes, too, and that's a `revert`. Otherwise the tree after `git revert --drop <commit>` would not have dropped the changes!

**Comment by LemmingAvalanche** [comment]
Isn’t all of this implemented now?

**Comment by dscho** [comment]
> Isn’t all of this implemented now?
> 

I might have missed it, is the `drop!` part implemented now?

**Comment by LemmingAvalanche** [comment]
@dscho Oh no sorry, I just tested (git version 2.46.2) and `git rebase -i --autosquash` does not translate `drop!` commits to the `drop` command in the todo editor.

Disregard that.

**Comment by dscho** [comment]
All good. Thank you for keeping better tabs on these here tickets than I manage.
